overlord/snapshotstate/backend/backend.go

```diff
@@ -369,3 +374,158 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
 
 	return nil
 }
+
+type exportMetadata struct {
+	Format int       `json:"format"`
+	Date   time.Time `json:"date"`
+	Files  []string  `json:"files"`
+}
+
+type SnapshotExport struct {
+	// open snapshot files
+	snapshotFiles []*os.File
+
+	// remember setID mostly for nicer errors
+	setID uint64
+
+	// cached size, needs to be calculated with CalculateSize
+	size int64
+}
+
+// NewSnapshotExport will return a SnapshotExport structure. It must be
+// Close()ed after use to avoid leaking file descriptors.
+func NewSnapshotExport(ctx context.Context, setID uint64) (se *SnapshotExport, err error) {
+	var snapshotFiles []*os.File
+
+	defer func() {
+		// cleanup any open FDs if anything goes wrong
+		if err != nil {
+			for _, f := range snapshotFiles {
+				f.Close()
+			}
+		}
+	}()
+
+	// Open all files first and keep the file descriptors
+	// open. The caller should have locked the state so that no
+	// delete/change snapshot operations can happen while the
+	// files are getting opened.
+	err = Iter(ctx, func(reader *Reader) error {
+		if reader.SetID == setID {
+			// dup() the file descriptor
+			fd, err := syscall.Dup(int(reader.Fd()))
+			if err != nil {
+				return fmt.Errorf("cannot dup %v", reader.Name())
+			}
+			f := os.NewFile(uintptr(fd), reader.Name())
+			if f == nil {
+				return fmt.Errorf("invalid fd %v for %v", fd, reader.Name())
```

Comment:
The error message should probably be something more like:

```suggestion
				return fmt.Errorf("cannot open file from descriptor %d", fd)
```

Please adjust to your liking. This also feels like an internal error.